src/v/security/gssapi_authenticator.cc

```diff
@@ -197,9 +202,16 @@ ss::future<result<bytes>> gssapi_authenticator::authenticate(bytes auth_bytes) {
 
     _state = res.state;
     if (_state == state::complete) {
-        _principal = co_await _worker.submit(
-          [this]() { return _impl->principal(); });
+        _principal = co_await _worker.submit([this]() {
+            auto principal = _impl->principal();
+            // Clear the gssapi members, as they're no longer required.
+            _impl->reset();
+            return principal;
+        });
+        // Clear the impl struct, as it's no longer required.
         _impl.reset();
+    } else if (_state == state::failed) {
+        co_await _worker.submit([this]() { _impl->reset(); });
```

Comment:
Any reason to keep around `_impl` when we're in the failed state?

Reply:
The connection shall be torn down, so it's not really kept around.